src/mewbot/api/type_guard.py

```diff
@@ -0,0 +1,97 @@
+# SPDX-FileCopyrightText: 2023 Mewbot Developers <mewbot@quicksilver.london>
+#
+# SPDX-License-Identifier: BSD-2-Clause
+
+"""
+Type guard exists to provide methods for run time typing validation.
+
+E.g. a decorator which checks that an InputEvent being passed to a function is of appropriate
+     type.
+
+The intent of the tools here is to provide a more elegant alternative to constructs like
+
+.. code-block:: python
+
+    def matches(self, event: InputEvent) -> bool:
+        if not isinstance(event, DiscordMessageCreationEvent):
+            return False
+        [Function body]
+
+which could be found in a Trigger.
+
+Instead, using one of the decorators provided in this module, you might write
+
+
+.. code-block:: python
+
+    @check_input_event_against_sig_types
+    def matches(self, event: InputEvent) -> bool:
+        [Function body]
+
+Or, as a more full example
+
+.. code-block:: python
+
+    # Demo Code
+
+    from mewbot.io.discord import DiscordInputEvent, DiscordUserJoinInputEvent
+    from mewbot.io.http import IncomingWebhookEvent
+
+    class A(Trigger):
+
+        @staticmethod
+        def consumes_inputs() -> set[type[InputEvent]]:
+            return {DiscordInputEvent, IncomingWebhookEvent}
+
+        @check_input_event_against_sig_types
+        def matches(self, event: DiscordInputEvent | IncomingWebhookEvent) -> bool:
+            return True
+
+    print(A.matches) # <function A.matches at 0x7f6e703f4a40>
+    print(A.matches.__doc__) # <function A.matches at 0x7f6e703f4a40>
+    print(A().matches(InputEvent())) # False
+    print(A().matches(DiscordInputEvent())) # True
+    print(A().matches(DiscordUserJoinInputEvent("[some text]"))) # True
+
+"""
+
+import functools
+
+from typing import Callable, TypeVar, get_type_hints, Union
+
+from mewbot.api.v1 import Trigger, Condition, InputEvent
+
+Component = TypeVar("Component", bound=Union[Trigger, Condition])
+Event = TypeVar("Event", bound=InputEvent)
+
+
+def check_input_event_against_sig_types(
+    wrapped: Callable[[Component, Event], bool]
+) -> Callable[[Component, InputEvent], bool]:
+    """
+    Check an input event against the valid event types declared in the signature.
+
+    Introspects the function to determine the types of InputEvent should be passed to it.
+    Uses a decorator to check that the event being passed in is one of those.
+        - If it is, the function is run
+        - If it is not, False is returned
+    :param wrapped:
+    :return:
+    """
+    types = get_type_hints(wrapped)
+    if "event" not in types:
+        raise TypeError("Received function without 'event' parameter")
+
+    event_types: type[Event] = types["event"]
+    if not isinstance(event_types, (type, type(Union))):
```

Comment:
I also don't know if this is a valid transform from `UnionType`, but I think it might be? If the tests pass after my next comment, it should be fine.

Reply:
....

It was, as it turns out, not.

I have, however, hacked it. With _moderate_ python crimes.